### PR TITLE
Add support for logical types

### PR DESCRIPTION
**Pull request by tjwp**
This looks like a big set of changes, but a lot of the lines are README and verbose specs.

Summary from the CHANGELOG:

- Add support for generating schemas that contain logical types.
- Add support for `union`, `map`, and `array` methods to embed those types
  within other complex types.
- Add methods for primitives types (`string`, `int`, `long`, `float`, etc)
  that allow those types, including a logical type attribute, to be embedded
  within complex types.

The changes here rely on unreleased changes from https://github.com/apache/avro/pull/116

Prime: @jturkel

### CODE REVIEW

File: spec/avro/builder_spec.rb [hunk 1]
```diff
@@ -1029,6 +1027,166 @@
...
+
+  context "array of unions via method" do
+    subject(:schema_json) do
+      described_class.build do
+        record :unions do
+          required :a, array(union(:string, :int))
+        end
+      end
+    end
+    let(:expected) do
```
Comment (jturkel) [inline spec/avro/builder_spec.rb]: Is this `expected` ever used?

File: spec/avro/builder_spec.rb [hunk 2]
```diff
@@ -1029,6 +1027,166 @@
...
+
+  context "union field via method" do
+    subject(:schema_json) do
+      described_class.build do
+        record :array_or_map do
+          required :u, union(array(:int), map(:string))
+        end
+      end
+    end
+    let(:expected) do
```
Comment (jturkel) [inline spec/avro/builder_spec.rb]: Is this `expected` used?

File: lib/avro/builder/dsl.rb [hunk 3]
```diff
@@ -83,15 +84,12 @@ def cache
       end
 
       def create_named_type(name, avro_type_name, options = {}, &block)
-        create_and_configure_builtin_type(avro_type_name,
-                                          cache: cache,
-                                          internal: { _name: name,
-                                                      namespace: namespace },
-                                          options: options,
-                                          &block).tap do |type|
-          type.validate!
```
Comment (jturkel) [inline lib/avro/builder/dsl.rb]: Why don't we need the `validate!` call anymore?

Comment (jturkel) [inline lib/avro/builder/dsl.rb]: Nevermind. Looks like it was pushed into `create_and_configure_builtin_type`

**Comment by jturkel** [comment]
A few small comments. :shipit: (although the Travis builds for this project seem to be having some issues).

File: spec/avro/builder_spec.rb (hunk 1 above)
Comment (tjwp) [inline spec/avro/builder_spec.rb]: No, I decided not to support this syntax, so it can be removed. Same for the other example you flagged.

**Comment by tjwp** [comment]
Re: Travis, this relies on some of the changes that I made to the Avro gem. I still need to figure out the best way to make that available publicly.

**Pull request merged**